ldaptive/src/main/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDao.java

```diff
@@ -127,43 +170,76 @@ public void setGroupValueTranscoder(@Nullable ValueTranscoder<Group> transcoder)
             SearchOperation search = new SearchOperation(conn);
             final SearchRequest request = new SearchRequest(this.baseSearchDn, filter);
 
-            // calculate the actual limit based on the provided limit & maxSearchResults
-            final int actualLimit = limit == SEARCH_NO_LIMIT ? maxSearchResults : maxSearchResults == SEARCH_NO_LIMIT
-                    ? limit : limit > maxSearchResults ? maxSearchResults : limit;
-            if (actualLimit != SEARCH_NO_LIMIT) {
-                request.setControls(new PagedResultsControl(actualLimit));
+            // calculate the page size based on the provided limit & maxPageSize
+            int pageSize = maxPageSize;
+            if (limit != SEARCH_NO_LIMIT && pageSize > limit) {
+                pageSize = limit;
+            }
+            if (restrictMaxAllowedResults && maxPageSize != SEARCH_NO_LIMIT && pageSize > maxPageSize + 1) {
+                pageSize = maxPageSize + 1;
             }
+            final PagedResultsControl prc = new PagedResultsControl(pageSize);
+            request.setControls(prc);
+
+            // retrieve results
+            byte[] cookie = null;
+            int processed = 0;
+            do {
+                // execute request
+                prc.setCookie(cookie);
+                cookie = null;
+                final Response<SearchResult> response = search.execute(request);
+
+                // process response
+                final SearchResult result = response.getResult();
+                final Iterator<LdapEntry> entries = result.getEntries().iterator();
+                while ((limit == SEARCH_NO_LIMIT || processed < limit) && (!restrictMaxAllowedResults ||
+                        maxSearchResults == SEARCH_NO_LIMIT || processed < maxSearchResults) && entries.hasNext()) {
+                    final User user = new User();
+                    userMapper.map(entries.next(), user);
+                    if (users instanceof BlockingQueue) {
```

Comment:
It would be nice, I think, if the contract for `BlockingQueue.add(E e)` followed that for `BlockingQueue.put(E e)`, in that you would not have to ask the instance type.